Functionality to get assigns from socket and displaying more tailored examples for users

GoNZooo opened the pull request:
I've made this PR to make the examples displayed to users for the Live Assigns slightly more tailored to their assigns.

The code added will allow for an assign to be fetched from their socket so that it can be displayed in the example and run without them having to edit the already given example. The `Live` example will be omitted if an example cannot be found.

I also modified the `set_style` example as it seemed like this one was outdated/wrong.

Apologies if there is a specific style guide or something that I've not found. You also mentioned on ElixirForum that the Drab.Live module was where a function for getting the assigns was appropriate, so that's where I put it.

Feel free to aggressively refactor this to your taste or give feedback so I can make it fit.

GoNZooo commented:
Apologies if there is a specific style guide or something that I've not found. You also mentioned on ElixirForum that the `Drab.Live` module was where a function for getting the assigns was appropriate, so that's where I put it.

Feel free to aggressively refactor this to your taste or give feedback so I can make it fit.

grych commented on lib/drab/channel.ex:
```diff
@@ -70,19 +70,38 @@ defmodule Drab.Channel do
+      live_example =
+        # if there are no assigns for the current socket, don't display `Live` example
+        if assign_example do
+          %{Drab.Live => "socket |> poke(#{assign_example}: 42)"}
```
Let's change the assign value to a string, like \"forty two\". Some of the assign (eg properties) types may vary, so it is safer to show user a string.

grych commented on lib/drab/live.ex:
```diff
@@ -373,6 +373,18 @@ defmodule Drab.Live do
     end
   end
 
+  def socket_to_assigns(socket) do
```
This will be a public API function, so it needs a documentation. Also, the name could be better, what about just `assigns`? (in this case it should rename existing private function to something more meaningful, like `assigns_values` or `assigns_for_partial` - as it returns a map with assigns and values for a specific partial).

To be consistent with the API the function should be like below:

    def assigns(socket) - returns assigns for the index partial (take it from index/1)
    def assigns(socket, partial) - where partial is "show.html" etc
    def assigns(socket, view, partial) - for a specific view (atom)

The existing private `assigns/3` takes partial hash (internal) and the partial name (the exact name, "show.html", used only in exception description). For the new `assigns/3`, you can take the partial hash from `partial_hash(view, partial_name)`. For `assigns/2` and `assigns/1`, you will need a view name for the partial (or for the main partial) - grab it from `Drab.get_view(socket)`

GoNZooo commented:
I've now renamed it to `assigns` and it can be used with a combination of `socket`, `partial` and `view` as requested. I also made it return existing atoms as well as based it on the previous `assigns` function which I renamed to `assign_data_for_partial`.

I also added a line in the channel intro-text about being able to list the assigns for the current socket.

grych commented on lib/drab/channel.ex:
```diff
@@ -98,6 +116,10 @@ defmodule Drab.Channel do
 
           Examples:
       #{examples |> Enum.join("\n")}
+
```
IMO this is not necessary, examples should not be too long - you see them all the time socket reloads, may be annoying ;)
Also, not everyone is using `Drab.Live` - in this case this example may be confusing.

GoNZooo replied:
Do you think it could be an idea to add it to the info text conditionally when `assign_example` is non-nil?

grych replied:
Nope, it is not worth to add it to those examples at all. It would be way too much - it is just a teaser, they should not replace docs and tutorial :)

GoNZooo replied:
Allright, I've now removed it completely.

grych commented on lib/drab/live.ex:
```diff
@@ -373,6 +373,49 @@ defmodule Drab.Live do
   end
 
+  @doc """
+  Returns a list of the assigns for the current socket.
```
I would say "for the main partial"

grych commented on lib/drab/live.ex:
```diff
@@ -373,6 +373,49 @@ defmodule Drab.Live do
+
+  Examples:
+
+      iex> socket = Drab.get_socket(pid("0.546.0"))
```
Remove this line, it is not adding any value to the example.

grych commented on lib/drab/live.ex:
```diff
@@ -373,6 +373,49 @@ defmodule Drab.Live do
+
+  Examples:
+
+      iex> socket = Drab.get_socket(pid("0.546.0"))
```
As above

grych commented on lib/drab/live.ex:
```diff
@@ -373,6 +373,49 @@ defmodule Drab.Live do
+  @doc """
+  Like `assigns/2`, but returns the assigns for a given combination of a `view` and a `partial`.
+ 
+      iex> socket = Drab.get_socket(pid("0.546.0"))
```
As above.

grych commented:
Thanks a lot!

I also added a simple test, it is in `test/integration/live_test.exs:136`

GoNZooo commented:
Thanks for making my first contribution to `Drab` great! :)

grych commented:
Actually, it was few issues with this (tests are not covering iex message :))
First at all, it crashed while opening with iex with Drab.Live declared, but the current template was not .drab
Also, there is an issue with timings, sometimes system is not quick enough to update assigns (they are coming from the frontend) before showing the iex example messages. For those cases, I made default assign name (:text).

See 5cdb56e

GoNZooo commented:
Ah, right, I didn't think of this interaction. Do the commits seem to cover it or do we need to revert?

grych commented:
No worries, me too. The issue is that not all the parts are covered with tests, so it is not so easy to find out. But I still don't know how to "emulate" running in iex on tests.

Your changes are commited with my updates. Thanks!
